src/main/java/org/wildfly/security/keystore/KeyStoreWatcher.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.keystore;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Central point for watching for modifications to KeyStores.
+ *
+ * @author <a href="mailto:darran.lofthouse@jboss.com">Darran Lofthouse</a>
+ */
+class KeyStoreWatcher {
+
+    /*
+     * Some key points: - - Could have multiple key stores in the same folder. - Could have multiple key store instances
+     * interested in the same underlying store.
+     *
+     * To begin with lest just assume singleton, but maybe this could be replaced with a container specific version.
+     */
+
+
+    private final ExecutorService executor;
+    private final FileSystem fileSystem;
+
+    private volatile Map<Path, Map<String, List<Store>>> watchedPaths = new HashMap<Path, Map<String, List<Store>>>();
+    private final Map<Path, WatchKey> registrations = new HashMap<Path, WatchKey>();
+
+    private volatile WatchService watchService;
+
+    private static KeyStoreWatcher theWatcher = new KeyStoreWatcher();
+
+    private KeyStoreWatcher() {
+        fileSystem = FileSystems.getDefault();
+        executor = Executors.newCachedThreadPool();
```

Comment:
To me this was a big candidate as an absolute minimum to be something we obtain from the application server, the reason I used an Executor was so that the handling of the modifications could also be handled by the executor without blocking picking up further modifications.

Having said that, switching to a single thread to monitor the notifications and call modified could simplify this class, the KeyStoreSpi implementation could then decide if it performs the load synchronously or asynchronously (Which actually might be a better place for that decision).